src/transaction/log_replication_atomic.cpp

```diff
@@ -367,4 +370,26 @@ namespace cublog
 	  }
       }
   }
+
+  void
+  atomic_replicator::release_locks_by_tranid (cubthread::entry &thread_entry, const TRANID trid)
+  {
+    assert (m_locked_objects.count (trid) > 0);
+
+    for (auto it = m_locked_objects.lower_bound (trid); it != m_locked_objects.upper_bound (trid); it++)
+      {
+	lock_unlock_object (&thread_entry, & (it->second), oid_Root_class_oid, SCH_M_LOCK, true);
+      }
+
+    m_locked_objects.erase (trid);
+  }
+
+  void
+  atomic_replicator::acquire_lock (cubthread::entry &thread_entry, const TRANID trid, const OID *classoid)
+  {
+    lock_object (&thread_entry, classoid, oid_Root_class_oid, SCH_M_LOCK, LK_UNCOND_LOCK);
```

Comment:
I doubt we can skip the error checking here. If it fails, the isolation will be broken. 
If it's supposed to succeed, I think we should use assert_release (), and if not, we have to handle it in a way.

Reply:
And, the "kick-out user transactions" would come in here. We need a TODO comment for that.

Reply:
I strongly agree. While working on the replicator code, I've realized that error handling is challenging. Even if an error occurs during the process of applying logs, it continues to move forward. 
I've been contemplating whether to proceed as the existing code or not, but I did follow the existing code..; 

Perhaps cases where `lock_object()` fails are related to not being able to allocate memory internally. In such situations, would it be a good idea to implement a mechanism for retrying the lock?

Reply:
The code assumes that functions calling 'lock_object()' suppose to obtain the lock, and in case it fails to acquire the lock, transactions are aborted. However, it's a bit perplexing since there's no separate error handling in other functionalities. 

Currently, 'lock_object()' fails only when there are issues like memory allocation problems internally, and repeatedly attempting 'lock_object()' in such cases may not yield success. It might be a good idea to use `assert_release()` for now and explore options for error handling in the `replicator`. Determining how the replicator should behave in the event of an error during execution would likely make error handling much smoother.

Reply:
How about other return value? It has 7 return value? Are they not allowed at all?

Reply:
> Determining how the replicator should behave in the event of an error during execution would likely make error handling much smoother.

I think it's where @cristiarg should kick in a help. It's related to both the replicator and lock conflict.

Reply:
> How about other return values? It has 6 other return values. Are they not allowed at all?

Yes they are not allowed.
Except for LK_GRANTED, all the other return values are LK_NOT_GRANTED_* or currently unused.